tests/shouldwork/DDR/VendorDDR.hs

```diff
@@ -0,0 +1,335 @@
+  ( Signal slow ((D, D, D, D))
+  , Signal fast (D, D)
+  )
+
+topEntityAnn ::
+  Bool ->
+  String ->
+  TopEntity
+topEntityAnn withEna nm =
+  Synthesize
+    { t_name = nm
+    , t_inputs =
+        [ PortName "clk"
+        , PortName "rst"
+        ] P.++
+        (if withEna then [PortName "en"] else []) P.++
+        [ PortName "in_in"
+        , PortProduct "in_out"
+            [ PortName "1"
+            , PortName "2"
+            ]
+        ]
+    , t_output =
+        PortProduct "out"
+          [ PortProduct "in"
+              [ PortName "gen1"
+              , PortName "gen2"
+              , PortName "vendor1"
+              , PortName "vendor2"
+              ]
+          , PortProduct "out"
+              [ PortName "gen"
+              , PortName "vendor"
+              ]
+          ]
+    }
+
+topEntityEnaAnn, topEntityNoEnaAnn :: String -> TopEntity
+topEntityEnaAnn = topEntityAnn True
+topEntityNoEnaAnn = topEntityAnn False
+
+type TestBenchT slow fast =
+  ( "clkSlow" ::: Clock slow
+  , "clkFast" ::: Clock fast
+  , "en" ::: Enable slow
+  , "in_in" ::: Signal fast D
+  , "out_in" ::: Signal slow
+      ( "1" ::: D
+      , "2" ::: D
+      )
+  , "in_out" ::: Signal slow
+      ( "gen1" ::: D
+      , "gen2" ::: D
+      , "vendor1" ::: D
+      , "vendor2" ::: D
+      )
+  , "out_out" ::: Signal fast
+      ( "gen" ::: D
+      , "vendor" ::: D
+      )
+  , "done" ::: Signal fast Bool
+  )
+
+topEntityGeneric ::
+  DomCxt slow fast fPeriod reset =>
+  VendorIn slow fast ->
+  VendorOut slow fast ->
+  TopEntityEna slow fast
+topEntityGeneric vendorIn vendorOut clk rst en inIn outIn =
+  ( bundle (genInOut1, genInOut2, vendorInOut1, vendorInOut2)
+  , bundle (genOutOut, vendorOutOut)
+  )
+ where
+  (genInOut1, genInOut2) = unbundle $ ddrIn clk rst en (0, 0, 0) inIn
+  (vendorInOut1, vendorInOut2) = unbundle $ vendorIn clk rst en inIn
+  genOutOut = ddrOut clk rst en 0 outIn
+  vendorOutOut = vendorOut clk rst en outIn
+
+noEnable ::
+  TopEntityNoEna slow fast ->
+  TopEntityEna slow fast
+noEnable top = top0
+ where
+  top0 clk rst _en = top clk rst
+
+expOut :: [D]
+expOut = 0 : P.concatMap (\e -> [e + 100, e + 200]) [0 .. natToNum @TestLen - 2]
+
+genOutSyncReset :: [D] -> [D]
+genOutSyncReset es =
+  let
+    es0 = P.take 51 es P.++ P.replicate 6 0 P.++ P.drop 57 es
+  in P.take 13 es0 P.++ P.replicate 6 0 P.++ P.drop 19 es0
+
+
+xilinxOutSyncReset :: [D] -> [D]
+xilinxOutSyncReset es =
+  let
+    es0 = P.take 51 es P.++ P.replicate 6 0 P.++ P.drop 57 es
+  in P.take 12 es0 P.++ P.replicate 7 0 P.++ P.drop 19 es0
+
+outAsyncReset :: [D] -> [D]
+outAsyncReset es =
+  let
+    es0 = P.take 50 es P.++ P.replicate 7 0 P.++ P.drop 57 es
+  in P.take 11 es0 P.++ P.replicate 8 0 P.++ P.drop 19 es0
+
+genOutEnable :: [D] -> [D]
+genOutEnable es =
+  let
+    es0 = P.take 69 es P.++ P.take 8 (cycle [133, 233]) P.++ P.drop 77 es
+  in P.take 31 es0 P.++ P.take 8 (cycle [114, 214]) P.++ P.drop 39 es0
```

Comment:
Wow that's a lot of magic numbers! How did they come to be?

Reply:
Because I definitely did not feel like writing this in a comprehensible way which would be _way_ much more work! :-D

They come from looking at waveforms, understanding them, agreeing with them, and stuffing them in a list of expected values here.

This kind of code would be inexcusible in `src/` but we have more instances of magic numbers in `tests/`.

Also, don't bother running the Haskell simulation against them. I still need to file some bugs that our simulation doesn't agree with HDL simulation. Although I have to say the Xilinx DDR primitive has extremely surprising timing regarding resets and enables, I actually wonder whether _their_ simulation matches hardware.

Reply:
(also, note that 69 plus 8 equals 77 and 31 plus 8 equals 39, these do not actually change the length of the list of samples, they overwrite samples that were affected by a deasserted enable)